### PR TITLE
Retain asset without data for RENDER_WORLD-only assets

### LINKED ISSUE #19737: Somehow prevent confusion caused by Assets being removed due to not having RenderAssetUsages::MAIN_WORLD
## What problem does this solve or what need does it fill?

One of the common causes of hard-to-diagnose issues i see people having is having the wrong RenderAssetUsages set. Most of the time the issue is silent, making it hard to figure out why things aren't working.

Examples:
- https://github.com/bevyengine/bevy/issues/19207 which is a more specific feature request that could be covered by this
- https://github.com/bevyengine/bevy/issues/19102
- https://github.com/bevyengine/bevy/issues/17987
- https://github.com/bevyengine/bevy/discussions/15172
- https://discord.com/channels/691052431525675048/1380027984823455816
- https://discord.com/channels/691052431525675048/937158127491633152/1380049315887841310
- https://discord.com/channels/691052431525675048/1382751438110851212
- https://discord.com/channels/691052431525675048/1385132782514344038

## What solution would you like?

#19207 is a good step, but it only covers one of many places where this issue could occur.
- only affects picking
- only affects Mesh

I think one solution could look like this:
Don't fully remove the handle from Assets in [extract_render_asset](https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/render_asset.rs#L287). Either leave a marker in its place, or store the handle separately. This would likely require a method `Assets::remove_with_reason` or similar. Then, in the various methods for getting the asset data, log out a warning if one of those handles has been requested. Probably should be `warn_once!` or maybe a smarter thing based on `track_caller` where its logged once for each caller.

## What alternative(s) have you considered?

And abundance of documentation and materials could also somewhat alleviate this, but thats far from ideal.
Another option could be to make Assets::get return a Result instead of a Option, allowing it to report a reason why it failed, though that assumes people don't just let-else the Err into oblivion.

## Additional context

none so far

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

when `RenderAssets` with `RenderAssetUsages::RENDER_WORLD` and without `RenderAssetUsages::MAIN_WORLD` are extracted, the asset is removed from the assets collection. this causes some issues:
- systems which rely on the asset, like picking with meshes, fail with "asset not found" errors which are unintuitive.
- loading the asset by path a second time results in the asset being reloaded from storage, re-extracted and re-transferred to gpu, replacing the existing asset
- knowledge about the asset state is lost, we cannot tell if an asset is already loaded with `AssetServer::get_handle`
- metadata (image size, e.g.) is no longer available for the asset

### extraction:
- add `take_gpu_data` to the `RenderAsset` trait. use it to pull the data out of the asset for transfer, and leave the empty asset in the collection. default implementation just `clone`s the asset.
- if the data has already been taken, ~~panic. this follows from modifying an asset after extraction, which is always a code error, so i think panic here makes sense~~ _log an error_

### Mesh/RenderMesh:
- make `Mesh::attributes` and `Mesh::indices` options
- take them on extraction
- `expect` operations which access or modify the vertex data or indices if it has been extracted. accessing the vertex data after extraction is always a code error. fixes #19737 by resulting in the error `Mesh has been extracted to RenderWorld. To access vertex attributes, the mesh must have RenderAssetUsages::MAIN_WORLD`
- provide `try_xxx` operations which allow users to handle the access error gracefully if required (no usages as part of this pr, but provided for future)
- compute the mesh `Aabb` when gpu data is taken and store the result. this allows extracted meshes to still use frustum culling (otherwise using multiple copies of an extracted mesh now panics as `compute_aabb` relied on vertex positions). there's a bit of a tradeoff here: users may not need the Aabb and we needlessly compute it. but i think users almost always do want them, and computing once (for extracted meshes) is cheaper than the alternative, keeping position data and computing a fresh `Aabb` every time the mesh is used on a new entity.

### Image/GpuImage:
images are a little more complex because the data can be deliberately `None` for render-targets / GPU-written textures where we only want an uninitialized gpu-side texture.
- take `Image::data` on extraction
- record on the resulting `GpuImage` whether any data was found initially
- on subsequent modifications with no data, panic if there was data previously

corner case / issue: when used with `RenderAssetBytesPerFrameLimiter` there may be no previous gpu asset if it is still queued pending upload due to the bandwidth limit. this can result in a modified image with initial data skipping the `had_data` check, resulting in a blank texture. i think this is sufficiently rare that it's not a real problem, users would still hit the panic if the asset is transferred in time and the problem/solution should be clear when they do hit it.

### ShaderStorageBuffer/GpuShaderStorageBuffer
follows the same pattern as Image/GpuImage:
- take `ShaderStorageBuffer::data` on extraction
- record on the resulting `GpuShaderStorageBuffer` whether any data was found initially
- on modifications with no data, panic if there was data previously

we don't have the queue issue here because `GpuShaderStorageBuffer` doesn't implement `byte_len` so we can't end up queueing them.

#### other RenderAssets
i didn't modify the other `RenderAsset` types (`GpuAutoExposureCompensationCurve`, `GpuLineGizmo`, `RenderWireframeMaterial`, `PreparedMaterial`, `PreparedMaterial2d`, `PreparedUiMaterial`) on the assumption that ~~cloning these is cheap enough anyway~~ _the asset usages are not exposed so we should never call `take_gpu_data`. the default implementation panics with a message directing users to implement the method if required_

## Testing

only really tested within my work project. i can add some explicit tests if required.

### CODE REVIEW

**Comment by hukasu** [comment]
I wonder if it would be better having an enum like this just to have some more information
```rust
enum RenderAssetData {
    Available(Vec<u8>),
    Unavailable,
    SentToRenderWorld
}
```

**Comment by robtfm** [comment]
maybe ... i tried to avoid changing the api of the main-world types (particularly `Image::data` which is pub) but maybe it's worthwhile

File: crates/bevy_asset/src/render_asset.rs [hunk 1]
```diff
@@ -4,12 +4,13 @@ use serde::{Deserialize, Serialize};
 bitflags::bitflags! {
     /// Defines where the asset will be used.
     ///
-    /// If an asset is set to the `RENDER_WORLD` but not the `MAIN_WORLD`, the asset will be
-    /// unloaded from the asset server once it's been extracted and prepared in the render world.
+    /// If an asset is set to the `RENDER_WORLD` but not the `MAIN_WORLD`, the asset data (pixel data,
+    /// mersh vertex data, etc) will be removed from the cpu-side asset once it's been extracted and prepared
```
Comment (jasmine-nominal) [inline crates/bevy_asset/src/render_asset.rs]:
```suggestion
    /// mesh vertex data, etc) will be removed from the cpu-side asset once it's been extracted and prepared
```

File: crates/bevy_mesh/src/mesh.rs [hunk 2]
```diff
@@ -389,19 +412,21 @@ impl Mesh {
     /// Retrieves the vertex `indices` of the mesh.
     #[inline]
     pub fn indices(&self) -> Option<&Indices> {
-        self.indices.as_ref()
+        self.indices.as_ref().expect("Mesh has been extracted to RenderWorld. To access vertex attributes, the mesh must have RenderAssetUsages::MAIN_WORLD").as_ref()
```
Comment (alice-i-cecile) [inline crates/bevy_mesh/src/mesh.rs]: Could we swap these methods to return `Result` and avoid the expects here?

Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs]: We could … it is a usage error though, rather than “something that can happen and you should deal with”, like result usually implies. 

I don’t feel strongly, will change if you prefer it.

File: crates/bevy_mesh/src/mesh.rs [hunk 3]
```diff
@@ -128,8 +128,8 @@ pub struct Mesh {
     /// Uses a [`BTreeMap`] because, unlike `HashMap`, it has a defined iteration order,
     /// which allows easy stable `VertexBuffers` (i.e. same buffer order)
     #[reflect(ignore, clone)]
-    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
-    indices: Option<Indices>,
+    attributes: Option<BTreeMap<MeshVertexAttributeId, MeshAttributeData>>,
+    indices: Option<Option<Indices>>,
```
Comment (andriyDev) [inline crates/bevy_mesh/src/mesh.rs]: Can we make this an enum instead? These options don't convey anything semantic and are just a hack to represent 3 possible states, so let's just accurately represent those three states!

File: crates/bevy_mesh/src/mesh.rs [hunk 4]
```diff
@@ -128,8 +128,8 @@ pub struct Mesh {
     /// Uses a [`BTreeMap`] because, unlike `HashMap`, it has a defined iteration order,
     /// which allows easy stable `VertexBuffers` (i.e. same buffer order)
     #[reflect(ignore, clone)]
-    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
-    indices: Option<Indices>,
+    attributes: Option<BTreeMap<MeshVertexAttributeId, MeshAttributeData>>,
```
Comment (andriyDev) [inline crates/bevy_mesh/src/mesh.rs]: This Option needs a comment explaining why it's an option. Something like "we use an option to indicate whether the data has been extracted into the render world".

File: crates/bevy_mesh/src/mesh.rs (hunk 2 above)
Comment (andriyDev) [reply crates/bevy_mesh/src/mesh.rs]: I don't think the intent is for a user to "deal with it", but rather "let's not crash the program because one asset has the wrong flags set". So I am in favour of making these Results.

File: crates/bevy_mesh/src/mesh.rs [hunk 5]
```diff
@@ -1362,6 +1410,33 @@ impl Mesh {
             })
         }
     }
+
+    pub fn take_gpu_data(&mut self) -> Option<Self> {
```
Comment (andriyDev) [inline crates/bevy_mesh/src/mesh.rs]: Doc comment please! Also linking to RenderAsset::take_gpu_data would be good.

File: crates/bevy_render/src/render_asset.rs [hunk 6]
```diff
@@ -86,6 +86,18 @@ pub trait RenderAsset: Send + Sync + 'static + Sized {
...
+
+    /// Make a copy of the asset to be moved to the `RenderWorld` / gpu. Heavy internal data (pixels, vertex attributes)
+    /// should be moved into the copy, leaving this asset with only metadata.
+    /// A null return value can be used to indicate that the asset has already been extracted, and should not
+    /// have been modified on the CPU side (as it cannot be transferred to GPU again).
+    /// The previous GPU asset is also provided, which can be used to check if the modification is valid.
+    fn take_gpu_data(
+        source: &mut Self::SourceAsset,
+        _previous_gpu_asset: Option<&Self>,
+    ) -> Option<Self::SourceAsset> {
```
Comment (andriyDev) [inline crates/bevy_render/src/render_asset.rs]: Should we even provide a default impl here? Cloning generally seems like the wrong behaviour to me, since I'd expect pretty much every impl to want to transfer the data rather than doing a possibly expensive clone.

File: crates/bevy_render/src/render_asset.rs [hunk 7]
```diff
@@ -86,6 +86,18 @@ pub trait RenderAsset: Send + Sync + 'static + Sized {
         _param: &mut SystemParamItem<Self::Param>,
     ) {
     }
+
+    /// Make a copy of the asset to be moved to the `RenderWorld` / gpu. Heavy internal data (pixels, vertex attributes)
+    /// should be moved into the copy, leaving this asset with only metadata.
+    /// A null return value can be used to indicate that the asset has already been extracted, and should not
```
Comment (andriyDev) [inline crates/bevy_render/src/render_asset.rs]:
```suggestion
    /// [`None`] can be used to indicate that the asset has already been extracted, and should not
```

File: crates/bevy_render/src/render_asset.rs (hunk 6 above)
Comment (andriyDev) [inline crates/bevy_render/src/render_asset.rs]: I'm not a fan of this Option. Can we make the return type `Result<Self::SourceAsset, InvalidModification>` or something? None isn't clear that it indicates a problem.

File: crates/bevy_mesh/src/mesh.rs [hunk 8]
```diff
@@ -389,19 +412,21 @@ impl Mesh {
     /// Retrieves the vertex `indices` of the mesh.
     #[inline]
     pub fn indices(&self) -> Option<&Indices> {
-        self.indices.as_ref()
+        self.indices.as_ref().expect("Mesh has been extracted to RenderWorld. To access vertex attributes, the mesh must have RenderAssetUsages::MAIN_WORLD").as_ref()
     }
 
     /// Retrieves the vertex `indices` of the mesh mutably.
     #[inline]
     pub fn indices_mut(&mut self) -> Option<&mut Indices> {
```
Comment (andriyDev) [inline crates/bevy_mesh/src/mesh.rs:757]: All these mutation methods essentially stop functioning when the asset is extracted. What about runtime generated (and mutated) assets? Is the intent that users will like fully replace the Mesh with like `*mesh_mut = Mesh::new()`?

If so, this probably needs a migration guide.

File: crates/bevy_render/src/texture/gpu_image.rs [hunk 9]
```diff
@@ -35,6 +36,20 @@ impl RenderAsset for GpuImage {
         image.asset_usage
     }
 
+    fn take_gpu_data(
+        source: &mut Self::SourceAsset,
+        previous_gpu_asset: Option<&Self>,
+    ) -> Option<Self::SourceAsset> {
+        let data = source.data.take();
+
+        let valid_upload = data.is_some() || previous_gpu_asset.is_none_or(|prev| !prev.had_data);
```
Comment (andriyDev) [inline crates/bevy_render/src/texture/gpu_image.rs:47]: Why can't we just reupload here? We're losing a capability here I think, since previously hot-reloading an image would just send the data to the GPU, which would unconditionally upload. Am I missing something here?

Could you also test that hot reloading still works? I believe the `alter_sprite` example is probably sufficient with the `file_watcher` feature (and then messing with the loaded sprite).

Comment (robtfm) [reply crates/bevy_render/src/texture/gpu_image.rs:47]: so the purpose of `RenderAssetUsages::RENDER_WORLD` is to reduce CPU memory by not keeping data on the CPU that is only required on the GPU. we clearly can't re-upload data from the CPU that we have chosen to purge.

as i described in the top comment, this check is specifically to catch an error which was not previously possible: where a user modifies some metadata on a `RENDER_WORLD`-only image after it has been transferred. that's not possible currently because currently the image gets removed from the CPU-side `Assets` collection so cannot be accidentally modified.

keeping access to it CPU side has the benefits i listed right at the top, but introduces this new way to break things, so i added this check to help anyone who does so accidentally. without this check, the image would be treated as an uninitialized texture -- for images with `data: None` we create an uninitialized GPU-texture which is much faster -- and the GPU pixel data would be zeros, which will never be what the user wants if the image had data when first transferred.

we're not losing any capabilities. default is `MAIN_WORLD | RENDER_WORLD`, in which case you can modify and (re)upload as much as you like because we keep the pixel data CPU side.

hot reloading still works for `MAIN_WORLD | RENDER_WORLD` and for `RENDER_WORLD`-only as reloading an asset reloads the data too.

File: crates/bevy_mesh/src/mesh.rs (hunk 8 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs:757]: the pattern would be either to fully initialize in one step, or to initialize with `RenderAssetUsages::MAIN_WORLD`, make any modifications, then when it's finished change the `usage to RenderAssetUsages::RENDER_WORLD`. after that no more modifications are possible since the data is only on GPU and we can't get it back. 

if users want to make continuous runtime modifications, then they should use `MAIN_WORLD || RENDER_WORLD`, or fully replacing will also work.

this PR doesn't change anything in those scenarios, modifying an asset with `RENDER_WORLD` and not `MAIN_WORLD` is not possible currently either since the asset gets removed from the `Assets` collection. i can write something if you like, but the usage doesn't change.

File: crates/bevy_render/src/render_asset.rs (hunk 6 above)
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs]: hm. i'm not sure the defaul impl gets called anywhere currently (i think the remaining RenderAsset types don't allow the user to set the usages), but then, what should we do here? panic in the trait impl or implement it explicitly for all the RenderAsset types with a panic or a clone, and require users to implement it explicitly as well even if they won't use it? 

that doesn't seem helpful to me.

Comment (andriyDev) [reply crates/bevy_render/src/render_asset.rs]: Yes, require users to implement the trait method. If they really are willing to tolerate cloning all their data, I think that's fine, but that should be an explicit choice from the impl.

File: crates/bevy_mesh/src/mesh.rs (hunk 2 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs]: making these return results will add literally hundreds of unwraps to the code base, e.g. there are 104 from primitive mesh builder functions calling `with_inserted_attribute` which are all guaranteed to succeed since they operate on meshes they've just constructed themselves.

perhaps i can add try_xxx variants, so that if we find situations where there are panics that we don't want we have a way to manage them in future.

File: crates/bevy_render/src/render_asset.rs (hunk 6 above)
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs]: i changed it so the default returns `AssetExtractionError::NoImplementation`

**Comment by robtfm** [comment]
- added MeshExtractableData enum to avoid the nasty nested options / make the intent clearer.
- added try_xxx functions to avoid 100s of unwraps in the cases where we know the access is valid (and reduce the breaking changes to the public api).
- changed the extract functions to return an error, and the default impl to return `AssetExtractionError::NoExtractionImplementation`.

i still need to do a pass through to fix up docs for all the copy/pasted functions, but the logic should be complete now i think.

File: crates/bevy_mesh/src/mesh.rs [hunk 10]
```diff
@@ -1362,23 +2135,102 @@ impl Mesh {
...
+        let morph_targets = self.morph_targets.extract()?;
+        #[cfg(feature = "morph")]
+        let morph_target_names = self.morph_target_names.extract()?;
+
+        // store the aabb extents as they cannot be computed after extraction
+        if let Some(MeshAttributeData {
+            values: VertexAttributeValues::Float32x3(position_values),
+            ..
+        }) = attributes.as_ref()?.get(&Self::ATTRIBUTE_POSITION.id)
+        {
```
Comment (greeble-dev) [inline crates/bevy_mesh/src/mesh.rs]:
```suggestion
        }) = attributes.as_ref()?.get(&Self::ATTRIBUTE_POSITION.id)
            && !position_values.is_empty()     
        {
```

Not 100% sure it's actually possible to make and extract a mesh with an empty position attribute, but seems reasonable to avoid a panic here just in case.

File: crates/bevy_render/src/mesh/mod.rs [hunk 11]
```diff
@@ -131,6 +133,15 @@ impl RenderAsset for RenderMesh {
...
 
+    fn take_gpu_data(
+        source: &mut Self::SourceAsset,
+        _previous_gpu_asset: Option<&Self>,
+    ) -> Result<Self::SourceAsset, AssetExtractionError> {
+        source
+            .take_gpu_data()
+            .map_err(|_| AssetExtractionError::AlreadyExtracted)
+    }
+
```
Comment (greeble-dev) [inline crates/bevy_render/src/mesh/mod.rs:144]: If `Mesh::take_gpu_data` returns `MeshAccessError::NotFound` (e.g. because the mesh has no attributes), then this logic will wrongly report `AlreadyExtracted`. I repro'd this by changing the `generate_custom_mesh` example to make a mesh with no attributes.

I'm not sure what the solution is - maybe `AssetExtractionError` should also have a discriminant for missing data.

**Review by greeble-dev — CHANGES_REQUESTED** [review]
Code seems reasonable - just got a couple of requests to avoid some corner cases. I also tested a variety of examples, and hacked in a few extra tests (including glTFs with `RENDER_WORLD` only, and making some examples deliberately trigger errors).

I don't like how the `Mesh` interface has ended up, but I'm not sure that's a problem with this PR. My hot take is that it's a problem with `Mesh` trying to be three things at once:

1. A finished, immutable asset that can be efficiently shared and rendered.
2. A way to incrementally build a mesh by adding/modifying attributes and suchlike.
3. A way to animate meshes by mutating an asset in-place and re-extracting.

I think the solution could be:

- Add a new `MeshBuilder` struct and move all the `Mesh` mutation methods to `MeshBuilder`.
  - This struct wouldn't have to deal with `MeshAccess` errors since it's entirely CPU side, so a load of `try_xxx` methods disappear.
  - A `finish` function would consume the builder and return a `Mesh`.
- Change `Mesh` to be immutable, except by `take_gpu_data`.
  - Animation that wants to mutate the mesh now has to create a fresh `Mesh` and overwrite the entire asset.

That gives as clear separation between 1) and 2), but makes 3) more awkward and inefficient. I'd argue that's ok in the short-term since it still basically works and mutating the asset is always going to be somewhat inefficient. In the long-term there could be a separate way of efficiently sending fine-grained mesh updates to the GPU.

The `Mesh`/`MeshBuilder` separation might also make compressed and derived data (like AABBs) more robust and efficient. So `MeshBuilder::finish` would automatically calculate AABBs and optionally apply compression.

EDIT: I should have said that I don't think this PR needs to explore stuff like `MeshBuilder`. What the PR does right now - make the `Mesh` interface more awkward in the short-term - seems like a reasonable trade-off for solving `RENDER_WORLD` issues now.

File: crates/bevy_mesh/src/mesh.rs (hunk 10 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs]: i'll change this to not bubble the error here, just fail to calc the aabb. then the only "?" error will be AlreadyExtracted. that should solve this and the below issue.

it's legitimate to extract a mesh with no positions, i guess, if you are doing some custom processing that doesn't use positions? maybe?

**Comment by robtfm** [comment]
it would be nice to have a MeshBuilder with infallible access methods. then all the Mesh accessors could be result-based without scattering 100s of unwraps about.

i don't think i want to do it as part of this PR though.

**Comment by greeble-dev** [comment]
Ah, my bad, I should have been clearer that I think this PR is fine as it is. It would have been simpler if `MeshBuilder` already existed, but c'est la vie.

File: crates/bevy_mesh/src/mesh.rs (hunk 10 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs]: sorry i misunderstood - i see now why this is required and added it

File: crates/bevy_mesh/src/mesh.rs [hunk 12]
```diff
@@ -150,6 +254,9 @@ pub struct Mesh {
     /// Does nothing if not used with `bevy_solari`, or if the mesh is not compatible
     /// with `bevy_solari` (see `bevy_solari`'s docs).
     pub enable_raytracing: bool,
+    /// Precomputed min and max extents of the mesh position data. Used mainly for constructing `Aabb`s for frustum culling.
+    /// This data will be set if/when a mesh is extracted to the GPU
+    pub final_aabb_extents: Option<(Vec3, Vec3)>,
```
Comment (beicause) [inline crates/bevy_mesh/src/mesh.rs]: I think this should be a `Option<Aabb3d>` which is more clearer than a tuple. And could it be private?

File: crates/bevy_camera/src/primitives.rs [hunk 13]
```diff
@@ -18,6 +18,11 @@ pub trait MeshAabb {
 
 impl MeshAabb for Mesh {
     fn compute_aabb(&self) -> Option<Aabb> {
+        if let Some(extents) = self.final_aabb_extents {
+            // use precomputed extents
+            return Some(Aabb::from_min_max(extents.0, extents.1));
+        }
+
         let Some(VertexAttributeValues::Float32x3(values)) =
             self.attribute(Mesh::ATTRIBUTE_POSITION)
```
Comment (greeble-dev) [inline crates/bevy_camera/src/primitives.rs]:
```suggestion
        let Ok(VertexAttributeValues::Float32x3(values)) =
            self.try_attribute(Mesh::ATTRIBUTE_POSITION)
```

Fixes a panic if the mesh has been extracted but had no position attributes, so `final_aabb_extents` was not set.

**Review by greeble-dev — CHANGES_REQUESTED** [review]
Noticed one more issue while testing meshes with missing attributes.

File: crates/bevy_mesh/src/mesh.rs (hunk 12 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs]: changed the type, thanks.

it's used in bevy_camera so can't be private unless we add an accessor ... that might be better since it would stop users setting it wrongly, but would prevent users who have a legitimate reason to modify it ... not sure if there is such a reason.

File: crates/bevy_camera/src/primitives.rs (hunk 13 above)
Comment (robtfm) [reply crates/bevy_camera/src/primitives.rs]: thanks, looks like i fixed that already on the other pr without pushing it back into this one ...

**Comment by beicause** [comment]
> Code seems reasonable - just got a couple of requests to avoid some corner cases. I also tested a variety of examples, and hacked in a few extra tests (including glTFs with `RENDER_WORLD` only, and making some examples deliberately trigger errors).
> 
> I don't like how the `Mesh` interface has ended up, but I'm not sure that's a problem with this PR. My hot take is that it's a problem with `Mesh` trying to be three things at once:
> 
> 1. A finished, immutable asset that can be efficiently shared and rendered.
> 2. A way to incrementally build a mesh by adding/modifying attributes and suchlike.
> 3. A way to animate meshes by mutating an asset in-place and re-extracting.
> 
> I think the solution could be:
> 
> * Add a new `MeshBuilder` struct and move all the `Mesh` mutation methods to `MeshBuilder`.
>   
>   * This struct wouldn't have to deal with `MeshAccess` errors since it's entirely CPU side, so a load of `try_xxx` methods disappear.
>   * A `finish` function would consume the builder and return a `Mesh`.
> * Change `Mesh` to be immutable, except by `take_gpu_data`.
>   
>   * Animation that wants to mutate the mesh now has to create a fresh `Mesh` and overwrite the entire asset.
> 
> That gives as clear separation between 1) and 2), but makes 3) more awkward and inefficient. I'd argue that's ok in the short-term since it still basically works and mutating the asset is always going to be somewhat inefficient. In the long-term there could be a separate way of efficiently sending fine-grained mesh updates to the GPU.
> 
> The `Mesh`/`MeshBuilder` separation might also make compressed and derived data (like AABBs) more robust and efficient. So `MeshBuilder::finish` would automatically calculate AABBs and optionally apply compression.

I totally agree with @greeble-dev.
Related to https://github.com/bevyengine/bevy/issues/17000. My quick thoughts are moving all the current `Mesh` methods to something like `InfallibleMesh`, too and making `Mesh` only contains packed vertex+indices data for GPU and necessary info about attributes layout, morph targets, aabb, etc.

So copying vertex and indices to MeshAllocator is fast and no longer needs repacking data. This opens a door to direct constructing Mesh and mutating Mesh in-place if you have the raw vertex and indices data and info and want to avoid repacking.

I think the current approach of extracting vertex attributes is suboptimal. And `Mesh` and `InfallibleMesh`  have many identical methods, making things confusing.

Edit: Oh, think again, I can see the reason to store attributes on the Mesh since `mesh_picking` or physics engine may need to read the Mesh’s vertex positions so we can’t just store interleaved vertex attributes. In that case, the above is less relevant and I think the current PR is fine.
See discussion in #21986

**Comment by beicause** [comment]
Regarding the Mesh interface, I‘d prefer to remove these `try_xxx` methods and let them panic. Adding a `is_extracted_to_render_world` method for checking is sufficient IMO, these `try_xxx` methods are rarely useful.

**Comment by robtfm** [comment]
> Regarding the Mesh interface, I‘d prefer to remove these `try_xxx` methods and let them panic. Adding a `is_extracted_to_render_world` method for checking is sufficient IMO, these `try_xxx` methods are rarely useful.

since the change to results was at request of @andriyDev (and @alice-i-cecile) i'll defer to them. i agree fwiw

File: crates/bevy_mesh/src/mesh.rs [hunk 14]
```diff
@@ -128,12 +232,12 @@ pub struct Mesh {
...
-    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
-    indices: Option<Indices>,
+    attributes: MeshExtractableData<BTreeMap<MeshVertexAttributeId, MeshAttributeData>>,
+    indices: MeshExtractableData<Indices>,
     #[cfg(feature = "morph")]
-    morph_targets: Option<Handle<Image>>,
+    morph_targets: MeshExtractableData<Handle<Image>>,
     #[cfg(feature = "morph")]
-    morph_target_names: Option<Vec<String>>,
+    morph_target_names: MeshExtractableData<Vec<String>>,
```
Comment (beicause) [inline crates/bevy_mesh/src/mesh.rs:240]: These extractable data can be consolidated. This ensures you always extract them together and that they are either all with data or without data  simultaneously. Like this:
```rs
struct MeshExtractableData {
    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
    indices: Option<Indices>,
    #[cfg(feature = "morph")]
    morph_targets: Option<Handle<Image>>,
    #[cfg(feature = "morph")]
    morph_target_names: Option<Vec<String>>,
}

struct Mesh {
    ...
    extractable_data: Option<MeshExtractableData>,
    ...
}
```

Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs:240]: could be. i don't think it improves anything but i can change it if others disagree.

**Review by IceSentry — APPROVED** [review]
The new `_option` function feel a bit weird but I don't really have an alternative solution. I mean it could either always return the option or always map the option to a result by it's not a blocker.

LGTM

**Comment by IceSentry** [comment]
And yeah, I prefer keeping the `try_` function that return results. We should generally try to avoid panics as much as possible. I've hit way too many unexpected panics in rendering code. Arguably, we shouldn't even have a panicky variant but that's a completely different discussion.

File: crates/bevy_mesh/src/mesh.rs [hunk 15]
```diff
@@ -29,6 +29,110 @@ use wgpu_types::{VertexAttribute, VertexFormat, VertexStepMode};
...
+#[derive(Debug, Clone, PartialEq, Reflect, Default)]
+enum MeshExtractableData<T> {
+    Data(T),
+    #[default]
+    NoData,
+    ExtractedToRenderWorld,
+}
+
+impl<T> MeshExtractableData<T> {
+    // get a reference to internal data. returns error if data has been extracted, or if no
```
Comment (andriyDev) [inline crates/bevy_mesh/src/mesh.rs:54]: We should make these all doc comments.

File: crates/bevy_render/src/render_asset.rs (hunk 6 above)
Comment (andriyDev) [reply crates/bevy_render/src/render_asset.rs]: I much prefer making users implement this explicitly - pretty much every render asset should want to handle this correctly I think, especially since the default is just breaking.

File: crates/bevy_render/src/render_asset.rs [hunk 16]
```diff
@@ -280,9 +302,17 @@ pub(crate) fn extract_render_asset<A: RenderAsset>(
                     let asset_usage = A::asset_usage(asset);
                     if asset_usage.contains(RenderAssetUsages::RENDER_WORLD) {
                         if asset_usage == RenderAssetUsages::RENDER_WORLD {
-                            if let Some(asset) = assets.remove(id) {
-                                extracted_assets.push((id, asset));
-                                added.insert(id);
+                            if let Some(asset) = assets.get_mut_untracked(id) {
```
Comment (andriyDev) [inline crates/bevy_render/src/render_asset.rs:305]: I'm a little worried that we're silently mutating the asset. Is this cause for concern?

Comment (IceSentry) [reply crates/bevy_render/src/render_asset.rs:305]: I would argue it's kinda part of the contract of using the render world asset usage but I can see why some people wouldn't like it. I'm not sure what the alternative is unless we can communicate more details about why an asset changed. If a user could react to this change right now there would be no way to know if it was a user event or the render world extraction. I guess if they look at the asset and see missing data they would know that's what hapenned.

File: crates/bevy_mesh/src/mesh.rs (hunk 15 above)
Comment (robtfm) [reply crates/bevy_mesh/src/mesh.rs:54]: do we typically use doc comments for private methods on private structs? i will change if so but i didn't think it was standard.

**Comment by robtfm** [comment]
>I much prefer making users implement this explicitly - pretty much every render asset should want to handle this correctly I think, especially since the default is just breaking.

there are several `RenderAsset`s that don't implement it and don't need to because their `RenderAssetUsages` are not exposed. it's not useful to make authors add it to types that don't support it, as i mentioned last time.

**Comment by beicause** [comment]
> And yeah, I prefer keeping the `try_` function that return results. We should generally try to avoid panics as much as possible. I've hit way too many unexpected panics in rendering code. Arguably, we shouldn't even have a panicky variant but that's a completely different discussion.

My concern is that if we add more extractable resources (like RawMesh, etc.), we will also follow a similar pattern of providing `try_xxx` functions, which feels really bad. Also, If the extractable data isn’t consolidated, we have to check each field individually to determine if a resource data is complete, it's inconvenient.

**Comment by robtfm** [comment]
>we will also follow a similar pattern of providing try_xxx functions

you've seen the `InfallibleMesh` pr i made on top of this, right? there it would still be an extra unwrap() on the base function, but wouldn't have any duplicated `try_` functions.

**Comment by beicause** [comment]
Another possible approach I consider is to consolidate the extractable data and access it through something similar to entry API, thereby avoiding multiple unwrap operations.

File: crates/bevy_render/src/render_asset.rs (hunk 16 above)
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs:305]: i used `get_mut` initially, but it ended up in a broken feedback loop since re-extraction is triggered by asset modifications (which is required behaviour for intentional user modifications), including this one.

the alternative would be to use a `get_mut` variant that fires a different event i guess. but the behaviour of extraction is predictable enough -- the data will be removed in the extract stage -- that i don't think it's warranted.

**Comment by beicause** [comment]
> > we will also follow a similar pattern of providing try_xxx functions
> 
> you've seen the `InfallibleMesh` pr i made on top of this, right? there it would still be an extra unwrap() on the base function, but wouldn't have any duplicated `try_` functions.

That is unrelated. I mean if more extractable assets are added in the future, or if crate authors want to add their extractable assets, should the interfaces also follow the `try_xxx`+`xxx` pattern? I think this is an area that could be improved.

File: crates/bevy_mesh/src/mesh.rs (hunk 15 above)
Comment (alice-i-cecile) [reply crates/bevy_mesh/src/mesh.rs:54]: We typically do, as we have documentation built for private items on dev-docs.bevy.org.
